tests/test.ra2.js

```diff
@@ -0,0 +1,348 @@
+            0x4d, 0x55, 0xc5, 0x4a, 0x7e, 0x0d, 0x4d, 0x9b,
+            0x84, 0xb4, 0x32, 0x2b, 0x4d, 0x8a, 0x34, 0x8d,
+            0xc8, 0xcf, 0x19, 0x3b, 0x64, 0x82, 0x27, 0x9e,
+            0xa7, 0x70, 0x2a, 0xc1, 0xb8, 0xf3, 0x6a, 0x3a,
+            0xf2, 0x75, 0x6e, 0x1d, 0xeb, 0xb6, 0x70, 0x7a,
+            0x15, 0x18, 0x38, 0x00, 0xb4, 0x4f, 0x55, 0xb5,
+            0xd8, 0x03, 0x4e, 0xb8, 0x53, 0xff, 0x80, 0x62,
+            0xf1, 0x9d, 0x27, 0xe8, 0x2a, 0x3d, 0x98, 0x19,
+            0x32, 0x09, 0x7e, 0x9a, 0xb0, 0xc7, 0x46, 0x23,
+            0x10, 0x85, 0x35, 0x00, 0x96, 0xce, 0xb3, 0x2c,
+            0x84, 0x8d, 0xf4, 0x9e, 0xa8, 0x42, 0x67, 0xed,
+            0x09, 0xa6, 0x09, 0x97, 0xb3, 0x64, 0x26, 0xfb,
+            0x71, 0x11, 0x9b, 0x3f, 0xbb, 0x57, 0xb8, 0x5b,
+            0x2e, 0xc5, 0x2d, 0x8c, 0x5c, 0xf7, 0xef, 0x27,
+            0x25, 0x88, 0x42, 0x45, 0x43, 0xa4, 0xe7, 0xde,
+            0xea, 0xf9, 0x15, 0x7b, 0x5d, 0x66, 0x24, 0xce,
+            0xf7, 0xc8, 0x2f, 0xc5, 0xc0, 0x3d, 0xcd, 0xf2,
+            0x62, 0xfc, 0x1a, 0x5e, 0xec, 0xff, 0xf1, 0x1b,
+            0xc8, 0xdb, 0xc1, 0x0f, 0x54, 0x66, 0x9e, 0xfd,
+            0x99, 0x9b, 0x23, 0x70, 0x62, 0x37, 0x80, 0xad,
+            0x91, 0x6b, 0x84, 0x85, 0x6a, 0x4c, 0x80, 0x9e,
+            0x60, 0x8a, 0x93, 0xa3, 0xc8, 0x8e, 0xc4, 0x4b,
+            0x4d, 0xb4, 0x8e, 0x3e, 0xaf, 0xce, 0xcd, 0x83,
+            0xe5, 0x21, 0x90, 0x95, 0x20, 0x3c, 0x82, 0xb4,
+            0x7c, 0xab, 0x63, 0x9c, 0xae, 0xc3, 0xc9, 0x71,
+            0x1a, 0xec, 0x34, 0x18, 0x47, 0xec, 0x5c, 0x4d,
+            0xed, 0x84,
+            // client hash
+            0x00, 0x14, 0x9c, 0x91, 0x9e, 0x76, 0xcf, 0x1e,
+            0x66, 0x87, 0x5e, 0x29, 0xf1, 0x13, 0x80, 0xea,
+            0x7d, 0xec, 0xae, 0xf9, 0x60, 0x01, 0xd3, 0x6f,
+            0xb7, 0x9e, 0xb2, 0xcd, 0x2d, 0xc8, 0xf8, 0x84,
+            0xb2, 0x9f, 0xc3, 0x7e, 0xb4, 0xbe,
+            // credentials
+            0x00, 0x08, 0x9d, 0xc8, 0x3a, 0xb8, 0x80, 0x4f,
+            0xe3, 0x52, 0xdb, 0x62, 0x9e, 0x97, 0x64, 0x82,
+            0xa8, 0xa1, 0x6b, 0x7e, 0x4d, 0x68, 0x8c, 0x29,
+            0x91, 0x38,
+        ]);
+        let sentData = new Uint8Array();
+        let sock;
+
+        sock = new Websock;
+        sock.open("ws://example.com");
+        sock.send = (data) => {
+            let res = new Uint8Array(sentData.length + data.length);
+            res.set(sentData);
+            res.set(data, sentData.length);
+            sentData = res;
+        };
+        const credentials = {
+            password: '123456',
+        };
+        const state = new RA2State(sock, () => {
+            expect(sentData).to.eql(sendData);
+            done();
+        }, () => credentials, (hasUsername, fingerprint) => {});
+        sock.on('message', () => {
+            state.poll();
+        });
+        state.negotiateRA2neAuthAsync();
+        sock._websocket._receiveData(receiveData);
+    });
+
```

Comment:
It is more interesting to test these things via the `RFB` object. There we can make sure events are fired properly and such things.